File: VB/ShowNewObjectExample.Module/NewObjectFromNavigationItemController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Text

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base

Imports DevExpress.ExpressApp.SystemModule
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.ExpressApp.Security


Namespace ShowNewObjectExample.Module
	Public Class CreateNewObjectFromNavigationController
		Inherits ShowNavigationItemController
		Private newController As NewObjectViewController
		Private createdDetailView As DetailView
		Private Const CreateNewObjectItemId As String = "CreateNewIssue"
		Private Const CreateNewObjectNavigationItemActiveKey As String = "CreationAllowed"
		Private Const CreateNewObjectNavigationItemDefaultPath As String = "Issues/" & CreateNewObjectItemId
		Public Sub New()
			TargetWindowType = WindowType.Main
		End Sub
		Protected Overrides Sub InitializeItems()
			MyBase.InitializeItems()
			Dim newNavigationItem As ChoiceActionItem = FindNewNavigationItem()
			If newNavigationItem IsNot Nothing Then
				newNavigationItem.Active(CreateNewObjectNavigationItemActiveKey) = SecuritySystem.IsGranted(New ObjectAccessPermission(GetType(DemoIssue), ObjectAccess.Create))
			End If
		End Sub
		Protected Overrides Sub ShowNavigationItem(ByVal e As SingleChoiceActionExecuteEventArgs)
			If (e.SelectedChoiceActionItem IsNot Nothing) AndAlso e.SelectedChoiceActionItem.Enabled.ResultValue AndAlso e.SelectedChoiceActionItem.Id = CreateNewObjectItemId Then
				Dim workFrame As Frame = Application.CreateFrame(TemplateContext.ApplicationWindow)
				workFrame.SetView(Application.CreateListView(Application.CreateObjectSpace(), GetType(DemoIssue), True))
				newController = workFrame.GetController(Of NewObjectViewController)()
				If newController IsNot Nothing Then
					Dim newObjectItem As ChoiceActionItem = FindNewObjectItem()
					If newObjectItem IsNot Nothing Then
						AddHandler newController.NewObjectAction.Executed, AddressOf NewObjectAction_Executed
						newController.NewObjectAction.DoExecute(newObjectItem)
						RemoveHandler newController.NewObjectAction.Executed, AddressOf NewObjectAction_Executed
						e.ShowViewParameters.TargetWindow = TargetWindow.Default
						e.ShowViewParameters.CreatedView = createdDetailView
						'Cancel the default processing for this navigation item.
						Return
					End If
				End If
			End If
			'Continue the default processing for other navigation items.
			MyBase.ShowNavigationItem(e)
		End Sub
		Private Function FindNewObjectItem() As ChoiceActionItem
			For Each item As ChoiceActionItem In newController.NewObjectAction.Items
				If item.Data Is GetType(DemoIssue) Then
					Return item
				End If
			Next item
			Return Nothing
		End Function
		Private Function FindNewNavigationItem() As ChoiceActionItem
			Return ShowNavigationItemAction.FindItemByIdPath(CreateNewObjectNavigationItemDefaultPath)
		End Function
		Private Sub NewObjectAction_Executed(ByVal sender As Object, ByVal e As ActionBaseEventArgs)
			createdDetailView = TryCast(e.ShowViewParameters.CreatedView, DetailView)
			'Cancel showing the default View by the NewObjectAction
			e.ShowViewParameters.CreatedView = Nothing
		End Sub
	End Class

End Namespace

File: CS/HowToCreateNewObjectViaNavigationControl.Module.Win/NewObjectFromNavigationItemController.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;

using DevExpress.ExpressApp.SystemModule;
using DevExpress.Persistent.BaseImpl;
using DevExpress.ExpressApp.Security;


namespace HowToCreateNewObjectViaNavigationControl.Module.Win {
   public partial class NewObjectFromNavigationItemController : ShowNavigationItemController {
      private View createdView;
      protected override void ShowNavigationItem(SingleChoiceActionExecuteEventArgs args) {
         base.ShowNavigationItem(args);
         if ((args.SelectedChoiceActionItem != null) && args.SelectedChoiceActionItem.Enabled) {
            if (args.SelectedChoiceActionItem.Id == "NewTask") {

               //Create a new Frame and set a new View for it
               Frame frame = Application.CreateFrame(TemplateContext.ApplicationWindow);
               ObjectSpace objectSpace = Application.CreateObjectSpace();
               CollectionSource collectionSource =
                  (CollectionSource)Application.CreateCollectionSource(objectSpace, typeof(Task), "");
               frame.SetView(Application.CreateListView(Application.GetListViewId(typeof(Task)), collectionSource, true));

               //Subscribe the NewObjectAction's Executed event
               NewObjectViewController controller = frame.GetController<NewObjectViewController>();
               controller.NewObjectAction.Executed += new EventHandler<ActionBaseEventArgs>(NewObjectAction_Executed);
               
               //Execute the NewObjectAction
               controller.NewObjectAction.DoExecute(null);
               //Show the NewObjectAction's View when clicking the "NewTask" navigation item
               args.ShowViewParameters.CreatedView = createdView;
            }
         }
      }
      protected override void OnActivated() {
         base.OnActivated();
         Window.ViewChanged += new EventHandler(Window_ViewChanged);
      }
      //Deactivate the SaveAndClose Action
      void Window_ViewChanged(object sender, EventArgs e) {
         if (Window.View != null) {
            if (Window.View.Id == "Task_DetailView") {
               DetailViewController detailViewController = Window.GetController<DetailViewController>();
               detailViewController.SaveAndCloseAction.Active.SetItemValue("TaskDetailViewInMainWindow", false);
            }
         }
      }
      void NewObjectAction_Executed(object sender, ActionBaseEventArgs e) {
         //Get the View to be set for the NewObjectAction "NewTask" item
         createdView = e.ShowViewParameters.CreatedView;
         //Cancel showing this View by the NewObjectAction
         e.ShowViewParameters.CreatedView = null;
      }
      protected override void InitializeItems() {
         base.InitializeItems();
         foreach (ChoiceActionItem group in ShowNavigationItemAction.Items) {
            foreach (ChoiceActionItem item in group.Items) {
               if (item.Id == "NewTask")
                  item.Active.SetItemValue("CreationAllowed", SecuritySystem.IsGranted(new
                     ObjectAccessPermission(typeof(Task), ObjectAccess.Create)));
            }
         }
      }
   }
}
